pylib/gyp/generator/make.py

```diff
@@ -2435,33 +2450,52 @@ def CalculateMakefilePath(build_file, base_name):
         makefile_path = os.path.join(
             options.toplevel_dir, options.generator_output, makefile_name
         )
-        srcdir = gyp.common.RelativePath(srcdir, options.generator_output)
+        srcdir = replace_sep(gyp.common.RelativePath(srcdir, options.generator_output))
         srcdir_prefix = "$(srcdir)/"
 
     flock_command = "flock"
     copy_archive_arguments = "-af"
     makedep_arguments = "-MMD"
+
+    # some linkers don't support --start-group/--end-group (e.g. wasm-ld)
+    link_commands = LINK_COMMANDS_LINUX
+    if gyp.common.CrossCompileRequested():
```

Comment:
I believe wasm-ld is the only one that doesn't support archive groups rather than if it is cross compiling. Emscripten linker supports the options as well.

Could we replace this condition with wasm-ld detection instead? `gyp.common.GetFlavor` may also be adapted to return `wasm32`, `wasm32-wasi`, or `emscripten` so that user doesn't have to define `OS` in their build scripts like https://github.com/toyobayashi/emnapi-node-gyp-test/blob/main/package.json#L10

Reply:
Done. But auto detect flavor seems doesn't work, it throws strange errors. https://github.com/toyobayashi/emnapi-node-gyp-test/actions/runs/8507038774.

> gyp.common.GetFlavor may also be adapted to return wasm32, wasm32-wasi, or emscripten so that user doesn't have to define OS

emnapi currently adds `--target` cflags according to `OS` (gyp disallow undefined variables). But auto detect can only distinguish emscripten and wasi, only when user specifies --target=wasm32 can wasi and wasm32 be further distinguished.

Reply:
seems not hit `OS == "emscripten"` condition on Linux and macOS. https://github.com/toyobayashi/emnapi-node-gyp-test/blob/18364bdd30cb232ba9c77344b7e4cae2da7f7c3d/binding.gyp#L14

Windows build failed: binding.target.mk:93: *** target pattern contains no '%'.  Stop.

Reply:
explicit `-f make-<flavor>` works. need figure out why emscripten flavor failed to detected and what happened on Windows.